rfcs/0000-import-from-derivation.md

```diff
@@ -0,0 +1,119 @@
+---
+
+# Summary
+[summary]: #summary
+
+Fix some issues with importing from derivations so it is nicer to use both manually and with CI.
+Once this RFC is implemented, hydra.nixos.org should allow importing from derivations.
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes the entire dependent graph cannot be known statically (before any building).
+Examples of this would be a package with its own Nix-based build system—the dependency must be downloaded first,
+or leveraging a language specific build tool (Cabal, Cargo, etc) to generate a build plan.
+
+Import-from-derivation and recursive Nix are two ways to achieve this dynamism—the interleaving of planning the build graph and executing it.
+Implementation-wise, however, they are completely orthogonal so there's no reason not to work on both.
+
+Currently there are some issues with import-from-derivation, however. Per @shlevy's read-only recursive Nix RFC:
+
+- *Import-from-derivation breaks dry-run evaluation and separation of evaluation-time from build-time.*
+
+  This is solved by simply making dry-run not build imported derivations.
+  Those imports instead become stuck terms, and the derivations blocking them are enqueued.
+  Moreover, the conceptual layering of the Nix language and the derivation builder is preserved:
+  While yes, in non-dry-run mode the evaluator will still need to invoke the builder, the builder need not know anything about evaluation.
+
+- *Import-from-derivation won't work if your expression-producing build needs to run on a different machine than your evaluating machine, unless you have distributed builds set up at evaluation time*
+
+  By not building during evaluation, we can use remote build machines just as usual.
+
+This RFC details the fixes to those problems.
+With these fixes made, hydra.nixos.org should change policy to allow importing from derivations.
+In turn, Nixpkgs can stop including auto-generated code, such as the dumps of Hackage.
+This will vastly decrease churn, decreasing growth of that repository.
+
+# Detailed design
+[design]: #detailed-design
+
+The fundamental change is rather than evaluation unconditionally triggering execution, evaluation will suspend yielding the execution needed to be done so it can make further progress on resumption.
+This is probably easier to explain in terms of the implementation steps needed.
+
+## Black-hole unwinding
+
+There's currently a bug with `nix-repl` where if computation is interrupted (say with `SIGINT`), and the same computation is begun again, one will get an infinite recursion error despite there being no infinite recursion.
+This is because the thunks stay black-holed—the indication that they are currently evaluated—despite the computation being aborted.
+To fix this, we want aborted computations to instead unwind their stack, un-black-holing any thunks back the way they were before.
+Mathematically, black-hole unwinding ensures that evaluation is idempotent.
```

Comment:
Seeing that that's just a bug fix, I might just open an issue. That issue can be linked from any tracking the RFC as a whole if it is accepted.